File: examples/python/inverse-dynamics.py
import numpy as np
import dynamics as dyn

# load an URDF file
model, geom_model = dyn.build_models_from_urdf("examples/descriptions/ur5_robot.urdf")

# Build a data frame associated with the model
data = model.create_data()

# sample a random joint configuration, joint velocities and accelerations
q = dyn.random_configuration(model)
v = np.random.rand(model.nv, 1)
a = np.random.rand(model.nv, 1)

# computes the inverse dynamics using Recursive Newton-Euler Algorithm (RNEA)
# tau = dyn.rnea(model, data, q, v, a)

# print("Joint torques: " + str(tau))

File: crates/dynamics-rs/examples/inverse_dynamics.rs
use dynamics_rs::prelude::*;

fn main() {
    let urdf_path = "./examples/descriptions/ur5/ur5_robot.urdf"; // Path to the URDF file
    let mesh_path = "./examples/descriptions/ur5/"; // Optional mesh path

    // Build models from the URDF file
    let (model, _coll_model, _viz_model) =
        build_models_from_urdf(urdf_path, Some(mesh_path)).expect("Failed to parse URDF file");

    // Generate a random configuration, velocity and acceleration
    let q = random_configuration(&model);
    let v = Configuration::from_element(model.nv, 1.0);
    let a = Configuration::from_element(model.nv, 1.0);

    // Create data structure for the model
    let mut data = model.create_data();

    // Compute inverse dynamics
    inverse_dynamics(&model, &mut data, &q, &v, &a, None)
        .expect("Failed to compute inverse dynamics");

    // Print the computed torques
    println!("Computed torques: {}", data.tau);
}
